File: 프로그래머스/3/43164. 여행경로/여행경로.kt
import java.util.ArrayDeque

class Solution {
    fun solution(tickets: Array<Array<String>>): Array<String> {
        // 1) 그래프 구성: from -> destinations
        val g = HashMap<String, MutableList<String>>()
        for (t in tickets) {
            val from = t[0]
            val to = t[1]
            g.getOrPut(from) { mutableListOf() }.add(to)
        }

        // 2) 내림차순 정렬 후, 마지막을 꺼내면(삭제) 사전순(오름차순) 우선 사용
        for (list in g.values) {
            list.sortDescending()
        }

        // 3) Hierholzer (스택)
        val st = ArrayDeque<String>()
        st.addLast("ICN")

        val route = ArrayList<String>(tickets.size + 1)

        while (st.isNotEmpty()) {
            val cur = st.last()
            val nexts = g[cur]
            if (nexts != null && nexts.isNotEmpty()) {
                val nxt = nexts.removeAt(nexts.lastIndex)
                st.addLast(nxt)
            } else {
                route.add(st.removeLast())
            }
        }

        route.reverse()
        return route.toTypedArray()
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.swift
import Foundation

func solution(_ tickets: [[String]]) -> [String] {
    // 1) 인접 리스트 구성
    var graph: [String: [String]] = [:]
    for t in tickets {
        graph[t[0], default: []].append(t[1])
    }

    // 2) 사전순으로 가장 앞 목적지를 꺼내야 하므로
    //    배열의 마지막에서 pop 하게끔 "내림차순" 정렬
    for (k, v) in graph {
        graph[k] = v.sorted(by: >)
    }

    // 3) Hierholzer (스택 기반)
    var stack: [String] = ["ICN"]
    var route: [String] = []

    while let cur = stack.last {
        if var nexts = graph[cur], !nexts.isEmpty {
            // 다음 목적지 하나 사용(pop)
            let nxt = nexts.removeLast()
            graph[cur] = nexts
            stack.append(nxt)
        } else {
            // 더 갈 곳 없으면 경로에 확정
            route.append(stack.removeLast())
        }
    }

    // route는 역순으로 쌓였으니 뒤집기
    return route.reversed()
}